README.md
PalDB
==========
[![Maven Central](https://maven-badges.herokuapp.com/maven-central/net.soundvibe/paldb/badge.svg)](https://maven-badges.herokuapp.com/maven-central/net.soundvibe/paldb)
[![Build Status](https://travis-ci.org/soundvibe/PalDB.svg)](https://travis-ci.org/soundvibe/PalDB)
[![codecov](https://codecov.io/gh/soundvibe/PalDB/branch/master/graph/badge.svg)](https://codecov.io/gh/soundvibe/PalDB)

PalDB is an embeddable write-once key-value store written in Java.

What is PalDB?
-------------------

PalDB is an embeddable persistent key-value store with very fast read performance and compact store size. PalDB stores are single binary files.

PalDB's JAR is only 65K and has a single dependency (snappy, which isn't mandatory). It's also very easy to use with just a few configuration parameters.

**This is separate evolution of original PalDB.**

Improvements from PalDB 1.0.2
-------------------
- New StoreRW interface for reading and writing. It is fully thread-safe and makes no sacrifices on performance.
- StoreReader is now fully thread-safe without any performance decrease.
- StoreReader and StoreWriter can be used with generics (StoreReader<K<V> and StoreWriter<K,V>)
- Typesafe ``PalDBConfigBuilder`` for easier configuration
- There are no limits on how many keys you can store
- Duplicates could be optionally allowed
- Bloom filters could be enabled for even better read performance in some cases
- Simplified and easier to write serializers

Performance
-----------

PalDB is significantly less complex than other embeddable key-value stores and therefore allows a compact storage format and very high throughput. PalDB is specifically optimized for fast read performance and compact store sizes. Performances can be compared to in-memory data structures such as Java collections (e.g. HashMap, HashSet) or other key-values stores (e.g. LevelDB, RocksDB).

Current benchmark on a 3.1Ghz Macbook Pro with 10M integer keys index shows an average performance of ~1.6M reads/s for a memory usage 6X less than using a traditional HashSet. That is 5X faster throughput compared to LevelDB (1.8) or RocksDB (4.0).

Results of a multithreaded throughput benchmark between PalDB and RocksDB (higher is better):

![throughput](throughput.jpg)

Memory usage benchmark between PalDB and a Java HashSet (lower is better):

![memory](memory.jpg)

What is it suitable for?
------------------------
PalDB can be used for caching [Kafka](https://kafka.apache.org/) compacted topics. It will perform key lookups very efficiently by using very little memory and storing database file on disk.

Side data can be defined as the extra read-only data needed by a process to do its job. For instance, a list of stopwords used by a natural language processing algorithm is side data. Machine learning models used in machine translation, content classification or spam detection are also side data. When this side data becomes large it can rapidly be a bottleneck for applications depending on them. PalDB aims to fill this gap.

PalDB can replace the usage of in-memory data structures to store this side data with comparable query performances and by using an order of magnitude less memory. It also greatly simplifies the code needed to operate this side data as PalDB stores are single binary files, manipulated with a very simple API (see below for examples).

Code samples
------------

API documentation can be found [here](http://linkedin.github.com/PalDB/doc/javadoc/index.html).
How to use RW store
```java
try (var palDB = PalDB.<Integer,String>createRW(new File("store.paldb"))) {
    try (var initializer = palDB.init()) {
        palDB.put(1213, "foo");
    }
    var value = palDB.get(1213);
}
```


How to write a store
```java
try (var writer = PalDB.<Integer,String>createWriter(new File("store.paldb"))) {
    writer.put(1213, "foo");
}
```

How to read a store
```java
try (var reader = PalDB.<Integer,String>createReader(new File("store.paldb"))) {
    String val = reader.get(1213);
}
```

How to iterate on a store
```java
try (var reader = PalDB.<Integer,String>createReader(new File("store.paldb"))) {
    try (var stream : reader.stream()) {
      stream.forEach(entry -> System.out.println(entry.getKey() + ": " + entry.getValue()));
    }
}
```

For Scala examples, see [here](https://gist.github.com/mbastian/9b9b49a4b96333da33ec) and [here](https://gist.github.com/mbastian/440a706f5e863bb65622).

Use it
------

PalDB is available on Maven Central, hence just add the following dependency:
```
<dependency>
    <groupId>net.soundvibe</groupId>
    <artifactId>paldb</artifactId>
    <version>2.1.3</version>
</dependency>
```
Scala SBT
```
libraryDependencies += "net.soundvibe" % "paldb" % "2.1.3"
```


Frequently asked questions
--------------------------

**Can you open a store for writing subsequent times?**

Yes, you can use `StoreRW` for that.

**Are duplicate keys allowed?**

Duplicates are not allowed by default. But it could be changed in writer configuration if needed.
Duplicates are automatically enabled when using StoreRW to have the same semantics as Java HashMap.

**Do keys have an order when iterating?**

No, like a hashtable PalDB stores have no order.

Build
-----

PalDB requires Java 13+ and maven. The target Java version is 13.

```bash
mvn build
```

Performance tests are run separately from the build
```bash
mvn clean test -Dtag=performance
```

Test
----

We use the JUnit5 framework for our unit tests. You can run them via the `mvn clean test` command.

Coverage
--------

Coverage is run using JaCoCo. You can run a report via `mvn jacoco:report`. The report will be generated in `paldb/build/reports/jacoco/test/html/`.

Advanced configuration
----------------------

Write parameters:

+ `load.factor`,  index load factor (double) [default: 0.75]
+ `compression.enabled`, enable compression (boolean) [default: false]
+ `bloom.filter.enabled`, enable bloom filter (boolean) [default: false]
+ `bloom.filter.error.factor`, bloom filter error rate  (double) [default: 0.01]
+ `duplicates.enabled`, allow duplicates  (boolean) [default: false]
+ `write.buffer.size`, Number of elements to hold in write buffer before compaction occurs  (int) [default: 100000]
+ `write.auto.flush.enabled`, Enable writer auto flush  (boolean) [default: true]

Read parameters:

+ `mmap.data.enabled`, enable memory mapping for data (boolean) [default: true]
+ `mmap.segment.size`, memory map segment size (bytes) [default: 1GB]

Configuration values are passed at init time. Example using fluent builder:
```java
var config = PalDBConfigBuilder.create()
                .withMemoryMapSegmentSize(512 * 1024 * 1024)
                .withMemoryMapDataEnabled(false)
                .withIndexLoadFactor(0.75)
                .withEnableCompression(true)
                .withEnableBloomFilter(true)
                .build();
StoreReader<String,String> reader = PalDB.createReader(new File("store.paldb"), config);
```

A few tips on how configuration can affect performance:

+ Disabling memory mapping will significantly reduce performance as disk seeks will be performed instead.
+ Enabling the bloom filter makes sense when you expect to miss finding some values. It will greatly increase read performance in this case.
+ Compression can be enabled when the store size is a concern and the values are large (e.g. a sparse matrix). By default, PalDB already uses a compact serialization. Snappy is used for compression.
+ StoreRW is not optimized for putting huge amounts of entries in short intervals after `init()` is completed. Keeping `write.buffer.size` large will reduce flush times significantly with a cost of increased memory usage.

Custom serializer
-----------------

PalDB is primarily optimized for Java primitives and arrays but supports adding custom serializers so arbitrary Java classes can be supported.

Serializers can be defined by implementing the `Serializer` interface and its methods. Here's an example which supports the `java.awt.Point` class:

```java
public class PointSerializer implements Serializer<Point> {

  private static final VarHandle INT_HANDLE = MethodHandles.byteArrayViewVarHandle(int[].class, ByteOrder.BIG_ENDIAN);
  
  @Override
  public byte[] write(Point input) {
    var buffer = new byte[8];
    INT_HANDLE.set(buffer, 0, input.x);
    INT_HANDLE.set(buffer, 4, input.y);
    return buffer;
  }

  @Override
  public Point read(byte[] bytes) {
    int x = (int)INT_HANDLE.get(bytes, 0);
    int y = (int)INT_HANDLE.get(bytes, 4);
    return new Point(x, y);
  }
}
```

The `write` method serializes the instance by returning `byte[]`. The `read` method deserializes from `byte[]` and creates new object instances.

Serializer implementation should be registered using the `Configuration`:

```java
var configuration = new Configuration<>(new ColorSerializer(), new PointSerializer());
```

Use cases
---------

At LinkedIn, PalDB is used in analytics workflows and machine-learning applications.

Its usage is especially popular in Hadoop workflows because memory is rare yet critical to speed things up. In this context, PalDB often enables map-side operations (e.g. join) which wouldn't be possible with classic in-memory data structures (e.g Java collections). For instance, a set of 35M member ids would only use ~290M of memory with PalDB versus ~1.8GB with a traditional Java HashSet. Moreover, as PalDB's store files are single binary files it is easy to package and use with Hadoop's distributed cache mechanism.

Machine-learning applications often have complex binary model files created in the `training` phase and used in the `scoring` phase. These two phases always happen at different times and often in different environments. For instance, the training phase happens on Hadoop or Spark and the scoring phase in a real-time service. PalDB makes this process easier and more efficient by reducing the need of large CSV files loaded in memory.

Limitations
-----------
+ PalDB is optimal in replacing the usage of large in-memory data storage but still use memory (off-heap, yet much less) to do its job. Disabling memory mapping and relying on seeks is possible but is not what PalDB has been optimized for.

Contributions
-----------

Any helpful feedback is more than welcome. This includes feature requests, bug reports, pull requests, constructive feedback, etc.

Copyright & License
-------------------

PalDB © 2019 Linas Naginionis. Licensed under the terms of the Apache License, Version 2.0.
PalDB © 2015 LinkedIn Corp. Licensed under the terms of the Apache License, Version 2.0.


CHANGELOG.md
Change Log
==========
Version 2.1.3
--------------------------
- Fixed possible deadlock when trying to update store while iterating

Version 2.1.1
--------------------------
- Added `remove(key)` to `StoreWriter<K,V>` and `StoreInitializer<K,V>`

Version 2.1.0
--------------------------
- New StoreRW interface for reading and writing
- Simplified serializers
- Configuration is now using generics as well.
- Small performance improvements

Version 2.0.1
--------------------------
- Fixed memory mapped files leak when closing the file

Version 2.0.0
--------------------------
- Bumped minimal java version to java 13
- StoreReader and StoreWriter implement AutoCloseable
- New config builder
- Reader and writer use generics
- Reader is now fully thread-safe
- Bloom filters could be used for better performance


Version 1.2.0 *(June 26th 2016)*
--------------------------

### Bugfixes

- Custom serializers with interfaces as types don't work (#25)
- Fix exception handling (#15)
- Fix concurrent database access with multiple instances (#26)

Version 1.1.0 *(January 8th 2016)*
--------------------------

### New features

- Implement a StoreReader.keys() method (#13)

### Bugfixes

- Fixes the return type of apis (#9)
- A Exception Using reader.iterable() (#16)

### Performance improvements

- Share slot read buffers (#14)

### Compatibility

The prefix bytes have been modified (#21) as they were confusing so exceptionally this version breaks store compatibility with 1.0.0. This shouldn't happen anymore in the future for minor versions.

Version 1.0.0 *(July 1st 2015)*
--------------------------
 *  First public release

src/main/java/com/linkedin/paldb/impl/StoreImpl.java
/*
* Copyright 2015 LinkedIn Corp. All rights reserved.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
* http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
*/

package com.linkedin.paldb.impl;

import com.linkedin.paldb.api.*;
import com.linkedin.paldb.api.StoreRW;
import com.linkedin.paldb.utils.FileUtils;
import org.slf4j.*;

import java.io.*;
import java.nio.file.Files;


/**
 * Static implementation factory.
 */
public final class StoreImpl {

  private static final Logger log = LoggerFactory.getLogger(StoreImpl.class);

  private StoreImpl() {
  }

  public static <K,V> StoreReader<K,V> createReader(File file, Configuration<K,V> config) {
    if (file == null || config == null) {
      throw new NullPointerException();
    }
    log.info("Initialize reader from file {}", file.getName());
    return new ReaderImpl<>(config, file);
  }

  public static <K,V> StoreReader<K,V> createReader(InputStream stream, Configuration<K,V> config) {
    if (stream == null || config == null) {
      throw new NullPointerException();
    }
    log.info("Initialize reader from stream, copying into temp folder");
    try {
      File file = FileUtils.copyIntoTempFile("paldbtempreader", stream);
      log.info("Copied stream into temp file {}", file.getName());
      return new ReaderImpl<>(config, file);
    } catch (IOException ex) {
      throw new UncheckedIOException(ex);
    }
  }

  public static <K,V> StoreWriter<K,V> createWriter(File file, Configuration<K,V> config) {
    if (file == null || config == null) {
      throw new NullPointerException();
    }
    try {
      log.info("Initialize writer from file {}", file.getName());
      Files.createDirectories(file.isDirectory() ? file.toPath() : file.toPath().toAbsolutePath().getParent());
      return new WriterImpl<>(config, file);
    } catch (IOException ex) {
      throw new UncheckedIOException(ex);
    }
  }

  public static <K,V> StoreWriter<K,V> createWriter(OutputStream stream, Configuration<K,V> config) {
    if (stream == null || config == null) {
      throw new NullPointerException();
    }
    log.info("Initialize writer from stream");
    return new WriterImpl<>(config, stream);
  }

  public static <V, K> StoreRW<K, V> createRW(File file, Configuration<K,V> config) {
    if (file == null || config == null) {
      throw new NullPointerException();
    }
    log.info("Initialize RW from file {}", file.getName());
    try {
        Files.createDirectories(file.isDirectory() ? file.toPath() : file.toPath().toAbsolutePath().getParent());
        return new StoreRWImpl<>(config, file);
    } catch (IOException e) {
        throw new UncheckedIOException(e);
    }
  }
}


src/main/java/com/linkedin/paldb/impl/StorageWriter.java
/*
* Copyright 2015 LinkedIn Corp. All rights reserved.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
* http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
*/

package com.linkedin.paldb.impl;

import com.linkedin.paldb.api.Configuration;
import com.linkedin.paldb.api.errors.*;
import com.linkedin.paldb.utils.*;
import org.slf4j.*;

import java.io.*;
import java.nio.*;
import java.nio.channels.FileChannel;
import java.nio.file.Files;
import java.text.DecimalFormat;
import java.util.*;

import static com.linkedin.paldb.impl.StorageSerialization.*;
import static com.linkedin.paldb.utils.DataInputOutput.*;

/**
 * Internal write implementation.
 */
public class StorageWriter {

  private static final Logger log = LoggerFactory.getLogger(StorageWriter.class);
  // Configuration
  private final Configuration<?,?> config;
  private final double loadFactor;
  // Output
  private final File tempFolder;
  private final OutputStream outputStream;
  // Index stream
  private File[] indexFiles;
  private DataOutputStream[] indexStreams;
  // Data stream
  private File[] dataFiles;
  private DataOutputStream[] dataStreams;
  // Cache last value
  private byte[][] lastValues;
  private int[] lastValuesLength;
  // Data length
  private long[] dataLengths;
  // Index length
  private long indexesLength;
  // Max offset length
  private int[] maxOffsetLengths;
  // Number of keys
  private long keyCount;
  private long[] keyCounts;
  private long[] actualKeyCounts;
  // Number of values
  private long valueCount;
  // Number of collisions
  private int collisions;

  private final long segmentSize;
  private final boolean duplicatesEnabled;

  StorageWriter(Configuration<?,?> configuration, OutputStream stream) {
    config = configuration;
    loadFactor = config.getDouble(Configuration.LOAD_FACTOR);
    if (loadFactor <= 0.0 || loadFactor >= 1.0) {
      throw new IllegalArgumentException("Illegal load factor = " + loadFactor + ", should be between 0.0 and 1.0.");
    }

    // Create temp path folder
    tempFolder = FileUtils.createTempDir("paldbtempwriter");
    tempFolder.deleteOnExit();
    log.info("Creating temporary folder at {}", tempFolder);
    outputStream = stream instanceof BufferedOutputStream ? stream : new BufferedOutputStream(stream);
    indexStreams = new DataOutputStream[0];
    dataStreams = new DataOutputStream[0];
    indexFiles = new File[0];
    dataFiles = new File[0];
    lastValues = new byte[0][];
    lastValuesLength = new int[0];
    dataLengths = new long[0];
    maxOffsetLengths = new int[0];
    keyCounts = new long[0];
    actualKeyCounts = new long[0];
    segmentSize = config.getLong(Configuration.MMAP_SEGMENT_SIZE);
    duplicatesEnabled = config.getBoolean(Configuration.ALLOW_DUPLICATES);
  }

  private static final byte[] REMOVED_BYTES = new byte[] {REMOVED_ID};

  public void put(byte[] key, byte[] value) throws IOException {
    int keyLength = key.length;

    //Get the Output stream for that keyLength, each key length has its own file
    DataOutputStream indexStream = getIndexStream(keyLength);

    // Write key
    indexStream.write(key);

    // Check if the value is identical to the last inserted
    byte[] lastValue = lastValues[keyLength];
    boolean sameValue = lastValue != null && Arrays.equals(value, lastValue);

    // Get data stream and length
    long dataLength = dataLengths[keyLength];
    if (sameValue) {
      dataLength -= lastValuesLength[keyLength];
    }

    // Write offset and record max offset length
    int offsetDataLength = LongPacker.packLong(indexStream, dataLength);
    maxOffsetLengths[keyLength] = Math.max(offsetDataLength, maxOffsetLengths[keyLength]);
    byte[] val;
    if (value == null) {
      val = REMOVED_BYTES;
      LongPacker.packInt(indexStream, REMOVED_ID);
    } else {
      val = value;
      LongPacker.packInt(indexStream, 0);
    }

    // Write if data is not the same
    if (!sameValue) {
      // Get stream
      DataOutputStream dataStream = getDataStream(keyLength);

      // Write size and value
      int valueSize = LongPacker.packInt(dataStream, val.length);
      dataStream.write(val);

      // Update data length
      dataLengths[keyLength] += valueSize + val.length;

      // Update last value
      lastValues[keyLength] = val;
      lastValuesLength[keyLength] = valueSize + val.length;

      valueCount++;
    }

    keyCount++;
    keyCounts[keyLength]++;
    actualKeyCounts[keyLength]++;
  }

  public void close() throws IOException {
    // Close the data and index streams
    for (var dos : dataStreams) {
      if (dos != null) {
        dos.close();
      }
    }

    for (var dos : indexStreams) {
      if (dos != null) {
        dos.close();
      }
    }

    // Stats
    log.info("Number of keys: {}", keyCount);
    log.info("Number of values: {}", valueCount);

    var bloomFilter = config.getBoolean(Configuration.BLOOM_FILTER_ENABLED) ?
            new BloomFilter(keyCount, config.getDouble(Configuration.BLOOM_FILTER_ERROR_FACTOR, 0.01)) :
            null;


    // Prepare files to merge
    List<File> filesToMerge = new ArrayList<>();
    try {
      // Build index file
      for (int i = 0; i < indexFiles.length; i++) {
        if (indexFiles[i] != null) {
          filesToMerge.add(buildIndex(i, bloomFilter));
        }
      }

      //Write metadata file
      File metadataFile = new File(tempFolder, "metadata.dat");
      metadataFile.deleteOnExit();
      try (var metadataOutputStream = new DataOutputStream(new BufferedOutputStream(new FileOutputStream(metadataFile)))) {
        writeMetadata(metadataOutputStream, bloomFilter);
      }

      filesToMerge.add(0, metadataFile);

      // Stats collisions
      log.info("Number of collisions: {}", collisions);

      // Add data files
      for (File dataFile : dataFiles) {
        if (dataFile != null) {
          filesToMerge.add(dataFile);
        }
      }

      // Merge and write to output
      checkFreeDiskSpace(filesToMerge);
      mergeFiles(filesToMerge, outputStream);
    } finally {
      outputStream.close();
      cleanup(filesToMerge);
    }
  }

  private void writeMetadata(DataOutputStream dataOutputStream, BloomFilter bloomFilter) throws IOException {
    //Write format version
    dataOutputStream.writeUTF(FormatVersion.getLatestVersion().name());

    //Write time
    dataOutputStream.writeLong(System.currentTimeMillis());

    //Prepare
    int keyLengthCount = getNumKeyCount();
    int maxKeyLength = keyCounts.length - 1;

    //Write size (number of keys)
    dataOutputStream.writeLong(keyCount);

    if (bloomFilter != null) {
      //write bloom filter bit size
      dataOutputStream.writeInt(bloomFilter.bitSize());
      //write bloom filter bits array size
      dataOutputStream.writeInt(bloomFilter.bits().length);
      //write bloom filter hash functions
      dataOutputStream.writeInt(bloomFilter.hashFunctions());
      //write bloom filter bits
      for (final long bit : bloomFilter.bits()) {
        dataOutputStream.writeLong(bit);
      }
    } else {
      dataOutputStream.writeInt(0);
      //write bloom filter long array size
      dataOutputStream.writeInt(0);
      //write bloom filter hash functions
      dataOutputStream.writeInt(0);
    }
    //Write the number of different key length
    dataOutputStream.writeInt(keyLengthCount);

    //Write the max value for keyLength
    dataOutputStream.writeInt(maxKeyLength);

    // For each keyLength
    long datasLength = 0L;
    for (int i = 0; i < keyCounts.length; i++) {
      if (keyCounts[i] > 0) {
        // Write the key length
        dataOutputStream.writeInt(i);

        // Write key count
        dataOutputStream.writeLong(keyCounts[i]);
        dataOutputStream.writeLong(actualKeyCounts[i]);

        // Write slot count
        long slots = Math.round(keyCounts[i] / loadFactor);
        dataOutputStream.writeLong(slots);

        // Write slot size
        int offsetLength = maxOffsetLengths[i];
        dataOutputStream.writeInt(i + offsetLength);

        // Write index offset
        dataOutputStream.writeLong(indexesLength);

        // Increment index length
        indexesLength += (i + offsetLength) * slots;

        // Write data length
        dataOutputStream.writeLong(datasLength);

        // Increment data length
        datasLength += dataLengths[i];
      }
    }

    //Write the position of the index and the data
    long indexOffset = (long) dataOutputStream.size() + (Long.SIZE / Byte.SIZE) + (Long.SIZE / Byte.SIZE);
    dataOutputStream.writeLong(indexOffset);
    //data offset
    dataOutputStream.writeLong(indexOffset + indexesLength);
  }

  private MappedByteBuffer[] initIndexBuffers(FileChannel channel, long indexSize) {
    int bufIdx = 0;
    int bufArraySize = (int) (indexSize / segmentSize) + ((indexSize % segmentSize != 0) ? 1 : 0);
    var result = new MappedByteBuffer[bufArraySize];
    try {
      for (long offset = 0; offset < indexSize; offset += segmentSize) {
        long remainingFileSize = indexSize - offset;
        long thisSegmentSize = Math.min(segmentSize, remainingFileSize);
        result[bufIdx++] = channel.map(FileChannel.MapMode.READ_WRITE, offset, thisSegmentSize);
      }
      return result;
    } catch (IOException e) {
      throw new UncheckedIOException(e);
    }
  }

  private File buildIndex(int keyLength, BloomFilter bloomFilter) throws IOException {
    long count = keyCounts[keyLength];
    long duplicateCount = 0L;
    long numSlots = Math.round(count / loadFactor);
    int offsetLength = maxOffsetLengths[keyLength];
    int slotSize = keyLength + offsetLength;
    // Init index
    var indexFile = new File(tempFolder, "index" + keyLength + ".dat");
    try (var indexAccessFile = new RandomAccessFile(indexFile, "rw")) {
      indexAccessFile.setLength(numSlots * slotSize);
      try (var indexChannel = indexAccessFile.getChannel()) {
        var indexBuffers = initIndexBuffers(indexChannel, indexAccessFile.length());
      // Init reading stream
        File tempIndexFile = indexFiles[keyLength];
        try (DataInputStream tempIndexStream = new DataInputStream(new BufferedInputStream(new FileInputStream(tempIndexFile)))) {
          byte[] keyBuffer = new byte[keyLength];
          byte[] slotBuffer = new byte[slotSize];
          byte[] offsetBuffer = new byte[offsetLength];

          // Read all keys
          for (long i = 0; i < count; i++) {
            // Read key
            tempIndexStream.readFully(keyBuffer);

            // Read offset
            long offsetData = LongPacker.unpackLong(tempIndexStream);
            int head = LongPacker.unpackInt(tempIndexStream);
            boolean isRemoved = head == REMOVED_ID;
            // Hash
            long hash = Murmur3.hash(keyBuffer);
            if (bloomFilter != null) {
              bloomFilter.add(keyBuffer);
            }
            boolean collision = false;
            for (long probe = 0; probe < count; probe++) {
              long slot = ((hash + probe) % numSlots);
              getFromBuffers(indexBuffers, slot * slotSize, slotBuffer, slotSize, segmentSize);
              long found = LongPacker.unpackLong(slotBuffer, keyLength);
              if (found == 0L) {

                if (isRemoved) {
                  duplicateCount++;
                  break;
                }
                // The spot is empty use it
                putIntoBuffers(indexBuffers, slot * slotSize, keyBuffer, keyBuffer.length, segmentSize);
                int pos = LongPacker.packLong(offsetBuffer, offsetData);
                putIntoBuffers(indexBuffers, (slot * slotSize) + keyBuffer.length, offsetBuffer, pos, segmentSize);
                break;
              } else {
                collision = true;
                // Check for duplicates
                if (isKey(slotBuffer, keyBuffer)) {
                  if (isRemoved || duplicatesEnabled) {
                    putIntoBuffers(indexBuffers, slot * slotSize, keyBuffer, keyBuffer.length, segmentSize);
                    int pos = LongPacker.packLong(offsetBuffer, offsetData);
                    putIntoBuffers(indexBuffers, (slot * slotSize) + keyBuffer.length, offsetBuffer, pos, segmentSize);
                    duplicateCount++;
                    if (isRemoved) duplicateCount++;
                    break;
                  } else {
                    throw new DuplicateKeyException(
                            String.format("A duplicate key has been found for key bytes %s", Arrays.toString(keyBuffer)));
                  }
                }
              }
            }

            if (collision) {
              collisions++;
            }
          }

          String msg = "  Max offset length: " + offsetLength + " bytes" +
                  "\n  Slot size: " + slotSize + " bytes";

          log.info("Built index file {} \n{}", indexFile.getName(), msg);

        } finally {
          unmap(indexBuffers);
          if (tempIndexFile.delete()) {
            log.info("Temporary index file {} has been deleted", tempIndexFile.getName());
          }
        }
      }
    } catch (Exception e) {
      Files.deleteIfExists(indexFile.toPath());
      throw e;
    }
    keyCount -= duplicateCount;
    actualKeyCounts[keyLength] -= duplicateCount;
    return indexFile;
  }

  private static boolean isKey(byte[] slotBuffer, byte[] key) {
    return Arrays.compare(slotBuffer, 0, key.length, key, 0, key.length) == 0;
  }

  //Fail if the size of the expected store file exceed 2/3rd of the free disk space
  private void checkFreeDiskSpace(List<File> inputFiles) {
    //Check for free space
    long usableSpace = 0;
    long totalSize = 0;
    for (File f : inputFiles) {
      if (f.exists()) {
        totalSize += f.length();
        usableSpace = f.getUsableSpace();
      }
    }
    log.info("Total expected store size is {} Mb",
        new DecimalFormat("#,##0.0").format(totalSize / (1024 * 1024)));
    log.info("Usable free space on the system is {} Mb",
        new DecimalFormat("#,##0.0").format(usableSpace / (1024 * 1024)));
    if (totalSize / (double) usableSpace >= 0.66) {
      throw new OutOfDiskSpace("Aborting because there isn't enough free disk space");
    }
  }

  //Merge files to the provided fileChannel
  private void mergeFiles(List<File> inputFiles, OutputStream outputStream) throws IOException {
    long startTime = System.nanoTime();
    //Merge files
    for (File f : inputFiles) {
      if (f.exists()) {
        log.info("Merging {} size={}", f.getName(), f.length());
        Files.copy(f.toPath(), outputStream);
      } else {
        log.info("Skip merging file {} because it doesn't exist", f.getName());
      }
    }
    log.debug("Time to merge {} s", ((System.nanoTime() - startTime) / 1000000000.0));
  }

  private void cleanup(List<File> inputFiles) {
    for (File f : inputFiles) {
      try {
        if (Files.deleteIfExists(f.toPath())) {
          log.info("Deleted temporary file {}", f.getName());
        }
      } catch (IOException e) {
        log.error("Cannot delete file " + f, e);
      }
    }

    if (FileUtils.deleteDirectory(tempFolder)) {
      log.info("Deleted temporary folder at {}", tempFolder.getAbsolutePath());
    }
  }

  //Get the data stream for the specified keyLength, create it if needed
  private DataOutputStream getDataStream(int keyLength)
      throws IOException {
    // Resize array if necessary
    if (dataStreams.length <= keyLength) {
      var copyOfDataStreams = Arrays.copyOf(dataStreams, keyLength + 1);
      Arrays.fill(dataStreams, null);
      dataStreams = null;
      dataStreams = copyOfDataStreams;
      dataFiles = Arrays.copyOf(dataFiles, keyLength + 1);
    }

    DataOutputStream dos = dataStreams[keyLength];
    if (dos == null) {
      File file = new File(tempFolder, "data" + keyLength + ".dat");
      file.deleteOnExit();
      dataFiles[keyLength] = file;

      dos = new DataOutputStream(new BufferedOutputStream(new FileOutputStream(file)));
      dataStreams[keyLength] = dos;

      // Write one byte so the zero offset is reserved
      dos.writeByte(0);
    }
    return dos;
  }

  //Get the index stream for the specified keyLength, create it if needed
  private DataOutputStream getIndexStream(int keyLength) throws IOException {
    // Resize array if necessary
    if (indexStreams.length <= keyLength) {
      var copyOfIndexStreams = Arrays.copyOf(indexStreams, keyLength + 1);
      Arrays.fill(indexStreams, null);
      indexStreams = null;
      indexStreams = copyOfIndexStreams;
      indexFiles = Arrays.copyOf(indexFiles, keyLength + 1);
      keyCounts = Arrays.copyOf(keyCounts, keyLength + 1);
      actualKeyCounts = Arrays.copyOf(actualKeyCounts, keyLength + 1);
      maxOffsetLengths = Arrays.copyOf(maxOffsetLengths, keyLength + 1);
      lastValues = Arrays.copyOf(lastValues, keyLength + 1);
      lastValuesLength = Arrays.copyOf(lastValuesLength, keyLength + 1);
      dataLengths = Arrays.copyOf(dataLengths, keyLength + 1);
    }

    // Get or create stream
    DataOutputStream dos = indexStreams[keyLength];
    if (dos == null) {
      File file = new File(tempFolder, "temp_index" + keyLength + ".dat");
      file.deleteOnExit();
      dos = new DataOutputStream(new BufferedOutputStream(new FileOutputStream(file)));
      indexFiles[keyLength] = file;
      indexStreams[keyLength] = dos;
      dataLengths[keyLength]++;
    }
    return dos;
  }

  private int getNumKeyCount() {
    int res = 0;
    for (final long count : keyCounts) {
      if (count != 0L) {
        res++;
      }
    }
    return res;
  }
}
